### DISCUSSION TITLE
Is there any way to sort the entities?

### DISCUSSION THREAD

**Discussion by kirinzer** [google/filament#3953] [Q&A]
I could set the priority to order the entities, but the range is limited(only cover [0,7]). Clearly, it is not the right way.
So I found the other.
`Builder& blendOrder(size_t primitiveIndex, uint16_t order)`
But this func looks like not work for me, nothing changes. it is confusing.
I use the OpenGL backend on the iOS platform.

**Comment by romainguy** [comment]
`blendOrder()` only works for entities that require blending (e.g. transparent objects). Could you describe what you are trying to do? Filament sorts the entities internally for performance and correctness already.

**Reply by kirinzer** [reply]
I try to implement a video editor, and I choose the entity as an abstract layer. That means I would have many layers(entities). So I need to control the render order exactly.

**Reply by romainguy** [reply]
Ah in that case you are going to use Ortho projection I assume? If so you can just set a different Z value for each layer and let distance sorting do the work for you.

**Reply by kirinzer** [reply]
Yeap, I use Ortho projection. Every layer has a matrix to control the scale between viewport and resource, also include the z value.
I set a series of increasing values, in the range of negative numbers. No effect.
Another strange phenomenon is that on the iOS platform, my entity drawing order seems to be that the entities added or created later will be inserted at the bottom, which is not as expected. Because under the Android platform, the one created first will be placed at the bottom, and the one created later will be overlaid on top.

**Reply by romainguy** [reply]
You can't rely on the order in which they are added since we sort by depth. How do you set the z values? Sorting is done on the bounding boxes transformed by the entity's world transform.

**Reply by kirinzer** [reply]
TransformManager, I set z values for selected entity depend on the `setTransform()`.

**Reply by romainguy** [reply]
Then check what your bounding boxes are. Btw drawing of opaque objects is front to back, so Z values closer to the camera will be first.

**Reply by kirinzer** [reply]
I am not sure how to check bounding boxes, and I create an entity like that below:
```
    RenderableManager::Builder(1)
    .material(0, matInstance)
    .geometry(0, TRIANGLES, vb, ib, 0, 6)
    .culling(false)
    .receiveShadows(false)
    .castShadows(false)
    .boundingBox(Box().set(Vec3f{-1,-1,-1}, Vec3f{1,1,1}))
    .build(engine, renderable);
```
Then set matrix for the entity:
```
    // tm is a TransformManager obj
    // originalMat is a Mat4, and z value is [0,-1], start from -1
    tm.setTransform(i, originalMat);
    // check  the  returned value, the z value is right
    Mat4 m = tm.getWorldTransform(i);
```
When next entity will be created:
```
    static float z = -1;
    originalMat = Matrix4();
    originalMat = originalMat.Translate(Vec3f{0,0,z});
    z = z + 0.1;
```

Am I missing something?

**Reply by kirinzer** [reply]
Find the real problem. My colleague had changed a container type in Scene.h
```
// modified code
    std::unordered_set<utils::Entity> mEntities;
// original code
    tsl::robin_set<utils::Entity> mEntities;
```
